### PR TITLE
Optimize complicated if-else in routing table.

### PR DESCRIPTION
**Pull request by howardhuanghua**
Optimize the complicated if-else logic in RoutingTable.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java [hunk 1]
```diff
@@ -472,19 +474,12 @@ public Builder updateNumberOfReplicas(final int numberOfReplicas, final String[]
...
-                    // now, add "empty" ones
-                    for (int i = 0; i < (numberOfReplicas - currentNumberOfReplicas); i++) {
+                int delta = Math.abs(numberOfReplicas - currentNumberOfReplicas);
+                for (int i = 0; i < delta; i++) {
+                    if (currentNumberOfReplicas < numberOfReplicas) {
                         builder.addReplica();
-                    }
-                } else if (currentNumberOfReplicas > numberOfReplicas) {
-                    int delta = currentNumberOfReplicas - numberOfReplicas;
-                    if (delta <= 0) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:482]: Maybe just remove this condition as it's obviously always `false` and keep the rest the same (maybe also inline `delta` like we have in the "add" case). Seems to me that makes the logic easiest to digest if we want to change it?

**Review by original-brownbear — COMMENTED** [review]
I think we can make this slighy simpler by dropping some dead code but I'm not sure about the approach taken here. Using the same loop for adding and removing replicas seems really confusing to be honest.

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java (hunk 1 above)
Comment (howardhuanghua) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:482]: @original-brownbear Thanks for checking, I think we could pre-compute the delta and seperate the add and remove operation. Please help to review the update.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:482]: I'm with @original-brownbear here, using `Math.abs(expr)` and then branching on the sign of `expr` looks weird to me. Let's avoid the use of `Math#abs` (and therefore the super-broad `@SuppressForbidden` annotation).

Comment (howardhuanghua) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:482]: OK. I removed `abs` method. Thank you @DaveCTurner .

**Comment by original-brownbear** [comment]
Good to test

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc (unrelated)

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Comment by original-brownbear** [comment]
@howardhuanghua could you please merge latest `master` into your branch, the BwC tests are broken because it's outdated I think.
Thanks!

**Comment by howardhuanghua** [comment]
@original-brownbear , I have merged the lastest master and updated this branch. Please help to re-trigger the test again. Thank you.

**Comment by original-brownbear** [comment]
Jenkins test this

**Review by original-brownbear — APPROVED** [review]
LGTM

**Pull request merged**